h2o-persist-gcs/src/main/java/water/persist/PersistGcs.java

```diff
@@ -225,4 +227,90 @@ public void close() throws IOException {
       }
     };
   }
+
+  @Override
+  public OutputStream create(String path, boolean overwrite) {
+    final GcsBlob gcsBlob = GcsBlob.of(path);
+    Log.debug("Creating: " + gcsBlob.getCanonical());
+    final WriteChannel writer = storage.create(gcsBlob.getBlobInfo()).writer();
+    return new OutputStream() {
+      @Override
+      public void write(int b) throws IOException {
+        ByteBuffer buffer = ByteBuffer.wrap(new byte[]{(byte) b});
+        writer.write(buffer);
+      }
+
+      @Override
+      public void write(byte[] b) throws IOException {
+        ByteBuffer buffer = ByteBuffer.wrap(b);
+        writer.write(buffer);
+      }
+
+      @Override
+      public void write(byte[] b, int off, int len) throws IOException {
+        ByteBuffer buffer = ByteBuffer.wrap(b, off, len);
+        writer.write(buffer);
+      }
+
+      @Override
+      public void close() throws IOException {
+        writer.close();
+      }
+    };
+  }
+
+  @Override
+  public boolean rename(String fromPath, String toPath) {
+    final BlobId fromBlob = GcsBlob.of(fromPath).getBlobId();
+    final BlobId toBlob = GcsBlob.of(toPath).getBlobId();
+
+    storage.get(fromBlob).copyTo(toBlob);
+    return storage.delete(fromBlob);
+  }
+
+  @Override
+  public boolean exists(String path) {
+    try {
+      final BlobId blob = GcsBlob.of(path).getBlobId();
+      return storage.get(blob).exists();
+    } catch (FSIOException e) {
+      final String bucketName = GcsBlob.removePrefix(path);
+      return storage.get(bucketName).exists();
+    } catch (NullPointerException npe) {
+      return false;
+    }
+  }
+
+  @Override
+  public boolean delete(String path) {
+    final BlobId blob = GcsBlob.of(path).getBlobId();
```

Comment:
What about cache ? Should we modify it too ? Or force to refresh it ?

Reply:
I suggest not to modify the cache. The cache is okay for "type ahead" kind of operation but it should not be used anywhere else in the code. If necessary it should be invalidated here.

Reply:
Invalidation has been added.